include/picongpu/particles/atomicPhysics2/initElectrons/InitElectronAs.hpp

```diff
@@ -0,0 +1,266 @@
+/* Copyright 2023 Brian Marre
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software you can redistribute it andor modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+//! @file init methods for macro electrons
+
+#pragma once
+
+#include "picongpu/simulation_defines.hpp"
+// need physicalConstants.param
+
+#include <pmacc/algorithms/math/PowerFunction.hpp>
+#include <pmacc/math/Matrix.hpp>
+
+#include <cstdint>
+
+namespace picongpu::particles::atomicPhysics2::initElectrons
+{
+    struct InitElectronAs
+    {
+        using Matrix_3x3 = pmacc::math::Matrix<float_64, pmacc::math::CT::UInt32<3u, 3u>>;
+        using MatrixVector = pmacc::math::Matrix<float_64, pmacc::math::CT::UInt32<3u, 1u>>;
+
+    private:
+        /** clone all add attributes that exist in both electron and ion species
+         *
+         * excludes:
+         *  - particleId, new particle --> new ID required, init by default
+         *  - multiMask, faster to set hard than copy, set in Kernel directly
+         *  - momentum, is mass dependent and therefore always changes
+         */
+        template<typename T_IonParticle, typename T_ElectronParticle>
+        HDINLINE static void cloneAdditionalAttributes(
+            T_IonParticle& ion,
+            // cannot be const even though we do not write to the ion
+            T_ElectronParticle& electron)
+        {
+            namespace partOp = pmacc::particles::operations;
+
+            auto targetElectronClone = partOp::deselect<pmacc::mp_list<multiMask, momentum>>(electron);
+
+            // otherwise this deselect will create incomplete type compile error
+            partOp::assign(targetElectronClone, partOp::deselect<particleId>(ion));
+        }
+
+        //! fill space components of a Lorentz boots matrix
+        HDINLINE static void fillLorentzMatrix(
+            MatrixVector beta,
+            float_64 gammaIonSystem,
+            float_64 normBetaSquared,
+            Matrix_3x3& lorentzMatrix)
+        {
+// general contributions
+#pragma unroll
+            for(uint32_t i = 0u; i < 3u; i++)
+            {
+#pragma unroll
+                for(uint32_t j = static_cast<uint32_t>(0u); j < static_cast<uint32_t>(3u); j++)
+                {
+                    lorentzMatrix.element(i, j) = (gammaIonSystem - 1.) * beta.element(i, static_cast<uint32_t>(0u))
+                        * beta.element(j, static_cast<uint32_t>(0u)) / normBetaSquared;
```

Comment:
division by zero in cases where `normBetaSquared`

Reply:
fixed